src/go/configgenerator/filter_generator.go

```diff
@@ -0,0 +1,758 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configgenerator
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/options"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/util"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/util/httppattern"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/ptypes"
+	anypb "github.com/golang/protobuf/ptypes/any"
+
+	sc "github.com/GoogleCloudPlatform/esp-v2/src/go/configinfo"
+	aupb "github.com/GoogleCloudPlatform/esp-v2/src/go/proto/api/envoy/v9/http/backend_auth"
+	bapb "github.com/GoogleCloudPlatform/esp-v2/src/go/proto/api/envoy/v9/http/backend_auth"
+	commonpb "github.com/GoogleCloudPlatform/esp-v2/src/go/proto/api/envoy/v9/http/common"
+	scpb "github.com/GoogleCloudPlatform/esp-v2/src/go/proto/api/envoy/v9/http/service_control"
+
+	corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	transcoderpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/grpc_json_transcoder/v3"
+	hcpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/health_check/v3"
+	jwtpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/jwt_authn/v3"
+	routerpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/router/v3"
+	hcmpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	durationpb "github.com/golang/protobuf/ptypes/duration"
+	emptypb "github.com/golang/protobuf/ptypes/empty"
+	wrapperspb "github.com/golang/protobuf/ptypes/wrappers"
+	confpb "google.golang.org/genproto/googleapis/api/serviceconfig"
+	smpb "google.golang.org/genproto/googleapis/api/servicemanagement/v1"
+)
+
+// A wrapper of filter generation logic.
+type FilterGenerator struct {
+	FilterName string
+	FilterGenFun
```

Comment:
Can we also associate the per-route config generation here? Each filter can have a function to generate its own per-route config. `PerRouteConfigGenFunc`

Reply:
SG

Reply:
Just think about it again. The `PerRouteConfigGenFunc` is tightly coupled in `FilterGenFun` while making filter config. Or we can make `PerRouteConfigGenFunc` under `FilterGenFun` after making it as object while I don't see too much improvement. WDYT?

Reply:
Also considering not all the filters are using per route config.

Reply:
For each filter (JWT, SC v2, CORS, etc.), we want all the filter-generation code to be under one object or folder.

You can modify `FilterGenFunc` to return a list of per-route configs as well, that's fine. If the list is empty, then the filter doesn't need per-route config.

Reply:
PTAL